tunnel/tun.go

```diff
@@ -1,40 +1,46 @@
 package tunnel
 
 import (
-	"errors"
-	"os"
+	"io"
 
 	"github.com/eycorsican/go-tun2socks/common/log"
 	_ "github.com/eycorsican/go-tun2socks/common/log/simple" // Import simple log for the side effect of making logs printable.
+	"golang.org/x/sys/unix"
 )
 
 const vpnMtu = 1500
 
-// MakeTunFile returns an os.File object from a TUN file descriptor `fd`.
-func MakeTunFile(fd int) (*os.File, error) {
-	if fd < 0 {
-		return nil, errors.New("Must provide a valid TUN file descriptor")
-	}
-	file := os.NewFile(uintptr(fd), "")
-	if file == nil {
-		return nil, errors.New("Failed to open TUN file descriptor")
-	}
-	return file, nil
+// TUNFile is an io.ReadWriter wrapping a TUN file descriptor (UNIX platforms only).
+// This is a substitute for os.NewFile that avoids taking ownership of the file.
+type TUNFile int
+
+func (f TUNFile) Read(buf []byte) (int, error) {
```

Comment:
Would it be possible to check whether the `fd` is valid so we can fail fast instead of on first read/write? For instance calling [fcntl](https://man7.org/linux/man-pages/man2/fcntl.2.html) on init?

Reply:
(obsolete)